### PR TITLE
[CELEBORN-900] Prefer to use jemalloc for memory allocation

### PR DESCRIPTION
**Pull request by mddxhj**
### What changes were proposed in this pull request?

Only the Dockfile needs to change in this pr.

### Why are the changes needed?

When deploying celeborn for flink on kubernetes, Introducing jemalloc can improve pod memory usage.

### Does this PR introduce _any_ user-facing change?
None

### How was this patch tested?
Maybe starting a production job to test the memory usage improvement is needed.

### CODE REVIEW

**Comment by waitinfuture** [comment]
cc @zwangsheng @pan3793 @RexXiong could you take a look at this PR?

File: docker/Dockerfile [hunk 1]
```diff
@@ -23,10 +23,11 @@ ARG celeborn_gid=10006
 
 ENV CELEBORN_HOME=/opt/celeborn
 ENV PATH=/opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/busybox:${CELEBORN_HOME}/sbin:${CELEBORN_HOME}/bin
+ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so
 
 RUN set -ex && \
     apt-get update && \
-    apt-get install -y bash tini busybox bind9-utils telnet net-tools procps krb5-user dnsutils && \
+    apt-get install -y bash tini busybox bind9-utils telnet net-tools procps krb5-user dnsutils gpg libsnappy1v5 gettext-base libjemalloc-dev && \
```
Comment (pan3793) [inline docker/Dockerfile:30]: why `gpg` `libsnappy1v5` `gettext-base` are required?

```
➜  ~ apt depends libjemalloc-dev
libjemalloc-dev
  Depends: libjemalloc2 (= 5.2.1-1ubuntu1)
  Suggests: binutils
➜  ~ apt depends libjemalloc2
libjemalloc2
  Depends: libc6 (>= 2.29)
  Depends: libgcc-s1 (>= 3.3)
  Depends: libstdc++6 (>= 4.4.0)
```

**Comment by pan3793** [comment]
@mddxhj please fill in the PR description with the necessary introduction on how it works.

File: docker/Dockerfile (hunk 1 above)
Comment (mddxhj) [reply docker/Dockerfile]: @pan3793 Sorry for this, there's no need for introducing these three libs, I've fixed it.

Comment (pan3793) [reply docker/Dockerfile]: seems the runtime library `libjemalloc2` is enough, as we don't need dev stuff

File: docker/Dockerfile [hunk 2]
```diff
@@ -35,6 +35,8 @@ RUN set -ex && \
     useradd  --uid=${celeborn_uid} --gid=${celeborn_gid} celeborn -d /home/celeborn -m && \
     mkdir -p ${CELEBORN_HOME}
 
+ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so
```
[the remark this answers is not in the document]
Comment (Myasuka) [reply docker/Dockerfile:38]: From my experience, I think we could use code to switch the environment via `uname -m` (we can refer to the [flink-docker code](https://github.com/apache/flink-docker/blob/45c6d230407d89aa83b0d170dd056d6868cf808e/1.17/scala_2.12-java8-ubuntu/docker-entrypoint.sh#L92C3-L109)). 

I think it make sense to make this option configurable and we shall enable this by default.

[the remark this answers is not in the document]
Comment (pan3793) [reply docker/Dockerfile:38]: `uname -m` would be a good idea, but does it work in ENV command? 

Also, is it only beneficial inside the container? if not, we'd better supply a general solution. e.g. port the jemalloc detection and enabling logic into the bootstrap shell script, and introduce an env var like CELEBORN_DISABLE_JEMALLOC to make it configurable.

Then, the behavior is consistent inside and outside of the container: 
- enable jemalloc by default when libjemalloc.so is available
- user can set CELEBORN_DISABLE_JEMALLOC=1 to explicitly disable the jemalloc usage

thought?

[the remark this answers is not in the document]
Comment (Myasuka) [reply docker/Dockerfile:38]: +1. I think your proposal is almost the same as what Flink does currently. The `uname -m` would be executed in the `docker-entrypoint.sh` instead of `Dockerfile` building.

[the remark this answers is not in the document]
Comment (pan3793) [reply docker/Dockerfile:38]: To be clear, what I mean "bootstrap shell script" is `sbin/load-celeborn-env.sh`(or maybe anther one), not `docker-entrypoint.sh`

[the remark this answers is not in the document]
Comment (Myasuka) [reply docker/Dockerfile:38]: Thanks for the clarification. Reconfigure the environment `LD_PRELOAD` in script would align the behavior. However, once running outside the container, the location of `libjemalloc.so` would be totally different in different versions of linux environment. To not extend this scope of this PR (mainly focus on the container environment), I think we would not pay too much efforts to find the location of  `libjemalloc.so`. The improvement could be done for others who runn Celeborn in non-container environments.

In a nutshell, I think moving the setting of `LD_PRELOAD` to environment related scripts would be a good idea, but we will only detect the `/usr/lib/$(uname -m)-linux-gnu/libjemalloc.so` location in this PR.

[the remark this answers is not in the document]
Comment (pan3793) [reply docker/Dockerfile:38]: We can simply consider the Ubuntu/CentOS in the detection logic, and allow the user to overwrite the `LD_PRELOAD` in `conf/celeborn-env.sh` in an arbitrary location.

[the remark this answers is not in the document]
Comment (pan3793) [reply docker/Dockerfile:38]: @mddxhj could you make such changes if you don't have further concerns?

File: docker/Dockerfile (hunk 1 above)
Comment (Myasuka) [reply docker/Dockerfile]: `libjemalloc-dev` could help [profiling the memory](https://github.com/jemalloc/jemalloc/wiki/Use-Case%3A-Heap-Profiling) to analysis the native memory leak problem. This is extremely helpful for complex OOMKilled problem.

Comment (pan3793) [reply docker/Dockerfile]: Thanks for information, let's keep `libjemalloc-dev` then

File: docker/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (mddxhj) [reply docker/Dockerfile:38]: @pan3793 I think there's no other concerns，I'll try to migrate LD_PRELOAD configuration to load-celeborn-env.sh。

**Comment by cfmcgrady** [comment]
I believe this PR isn't just about the "flink on k8s", but also aims to enhance memory allocation/utilization when deploying the Celeborn server on k8s.

**Comment by pan3793** [comment]
maybe "[CELEBORN-900] Prefer to use jemalloc for memory allocation"

File: sbin/load-celeborn-env.sh [hunk 3]
```diff
@@ -69,3 +69,25 @@ if [ "$CELEBORN_PID_DIR" = "" ]; then
   export CELEBORN_PID_DIR="${CELEBORN_HOME}/pids"
 fi
 
+# jemalloc memory allocator is enabled by default, you may set DISABLE_JEMALLOC to true if it's not as you wish.
+DISABLE_JEMALLOC="false"
+maybe_enable_jemalloc() {
+    if [ "${DISABLE_JEMALLOC:-false}" == "false" ]; then
```
Comment (pan3793) [inline sbin/load-celeborn-env.sh:75]: nit: style, two indents

File: sbin/load-celeborn-env.sh [hunk 4]
```diff
@@ -69,3 +69,25 @@ if [ "$CELEBORN_PID_DIR" = "" ]; then
   export CELEBORN_PID_DIR="${CELEBORN_HOME}/pids"
 fi
 
+# jemalloc memory allocator is enabled by default, you may set DISABLE_JEMALLOC to true if it's not as you wish.
+DISABLE_JEMALLOC="false"
```
Comment (pan3793) [inline sbin/load-celeborn-env.sh:73]: I suppose this declaration should be removed? otherwise, it will overwrite the definition in `celeborn-env.sh`

File: docker/Dockerfile [hunk 5]
```diff
@@ -26,15 +26,14 @@ ENV PATH=/opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
...
-    apt-get install -y bash tini busybox bind9-utils telnet net-tools procps krb5-user dnsutils && \
+    apt-get install -y bash tini busybox bind9-utils telnet net-tools procps krb5-user dnsutils libjemalloc-dev && \
     ln -snf /bin/bash /bin/sh && \
     rm -rf /var/cache/apt/* && \
     mkdir /opt/busybox && \
     busybox --install /opt/busybox && \
     groupadd --gid=${celeborn_gid} celeborn && \
     useradd  --uid=${celeborn_uid} --gid=${celeborn_gid} celeborn -d /home/celeborn -m && \
     mkdir -p ${CELEBORN_HOME}
-
```
Comment (pan3793) [inline docker/Dockerfile:37]: nit: unnecessary change.

File: docker/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (mddxhj) [reply docker/Dockerfile]: > @mddxhj could you make such changes if you don't have further concerns?

@pan3793 I have changed the implementation as mentioned above.

File: sbin/load-celeborn-env.sh (hunk 3 above)
Comment (pan3793) [reply sbin/load-celeborn-env.sh:75]: please adding the configurable env vars into the `celeborn-env.sh.template`

**Comment by pan3793** [comment]
@mddxhj thanks for updating, left some comments. and please update the PR title too.

File: docker/Dockerfile (hunk 5 above)
Comment (mddxhj) [reply docker/Dockerfile]: have fix it

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

File: conf/celeborn-env.sh.template [hunk 6]
```diff
@@ -31,4 +31,5 @@
 # CELEBORN_WORKER_JAVA_OPTS="-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
 # CELEBORN_MASTER_JAVA_OPTS="-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
 # CELEBORN_PID_DIR="$CELEBORN_HOME/pids"
-# CELEBORN_LOG_DIR="$CELEBORN_HOME/logs"
\ No newline at end of file
+# CELEBORN_LOG_DIR="$CELEBORN_HOME/logs"
+DISABLE_JEMALLOC="false"
```
Comment (pan3793) [inline conf/celeborn-env.sh.template:35]:
```suggestion
# DISABLE_JEMALLOC="false"
```

File: conf/celeborn-env.sh.template [hunk 7]
```diff
@@ -31,4 +31,5 @@
 # CELEBORN_WORKER_JAVA_OPTS="-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
 # CELEBORN_MASTER_JAVA_OPTS="-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
 # CELEBORN_PID_DIR="$CELEBORN_HOME/pids"
-# CELEBORN_LOG_DIR="$CELEBORN_HOME/logs"
\ No newline at end of file
+# CELEBORN_LOG_DIR="$CELEBORN_HOME/logs"
+# DISABLE_JEMALLOC="false"
```
Comment (pan3793) [inline conf/celeborn-env.sh.template:35]:
```suggestion
# DISABLE_JEMALLOC="false"

```

File: sbin/load-celeborn-env.sh [hunk 8]
```diff
@@ -69,3 +69,24 @@ if [ "$CELEBORN_PID_DIR" = "" ]; then
...
+maybe_enable_jemalloc() {
+  if [ "${DISABLE_JEMALLOC:-false}" == "false" ]; then
+    JEMALLOC_PATH="/usr/lib/$(uname -m)-linux-gnu/libjemalloc.so"
+    JEMALLOC_FALLBACK="/usr/lib/x86_64-linux-gnu/libjemalloc.so"
+    if [ -f "$JEMALLOC_PATH" ]; then
+      export LD_PRELOAD=$LD_PRELOAD:$JEMALLOC_PATH
+    elif [ -f "$JEMALLOC_FALLBACK" ]; then
+      export LD_PRELOAD=$LD_PRELOAD:$JEMALLOC_FALLBACK
+    else
+      if [ "$JEMALLOC_PATH" = "$JEMALLOC_FALLBACK" ]; then
```
Comment (pan3793) [inline sbin/load-celeborn-env.sh:82]:
```suggestion
      if [ "$JEMALLOC_PATH" == "$JEMALLOC_FALLBACK" ]; then
```

File: sbin/load-celeborn-env.sh [hunk 9]
```diff
@@ -69,3 +69,24 @@ if [ "$CELEBORN_PID_DIR" = "" ]; then
...
+  if [ "${DISABLE_JEMALLOC:-false}" == "false" ]; then
+    JEMALLOC_PATH="/usr/lib/$(uname -m)-linux-gnu/libjemalloc.so"
+    JEMALLOC_FALLBACK="/usr/lib/x86_64-linux-gnu/libjemalloc.so"
+    if [ -f "$JEMALLOC_PATH" ]; then
+      export LD_PRELOAD=$LD_PRELOAD:$JEMALLOC_PATH
+    elif [ -f "$JEMALLOC_FALLBACK" ]; then
+      export LD_PRELOAD=$LD_PRELOAD:$JEMALLOC_FALLBACK
+    else
+      if [ "$JEMALLOC_PATH" == "$JEMALLOC_FALLBACK" ]; then
+        MSG_PATH=$JEMALLOC_PATH
```
Comment (pan3793) [inline sbin/load-celeborn-env.sh:83]:
```suggestion
        MSG_PATH="$JEMALLOC_PATH"
```

File: sbin/load-celeborn-env.sh [hunk 10]
```diff
@@ -69,3 +69,24 @@ if [ "$CELEBORN_PID_DIR" = "" ]; then
...
 
+# jemalloc memory allocator is enabled by default, you may set DISABLE_JEMALLOC to true in celeborn-env.sh if it's not as you wish.
+maybe_enable_jemalloc() {
+  if [ "${DISABLE_JEMALLOC:-false}" == "false" ]; then
+    JEMALLOC_PATH="/usr/lib/$(uname -m)-linux-gnu/libjemalloc.so"
+    JEMALLOC_FALLBACK="/usr/lib/x86_64-linux-gnu/libjemalloc.so"
+    if [ -f "$JEMALLOC_PATH" ]; then
+      export LD_PRELOAD=$LD_PRELOAD:$JEMALLOC_PATH
+    elif [ -f "$JEMALLOC_FALLBACK" ]; then
+      export LD_PRELOAD=$LD_PRELOAD:$JEMALLOC_FALLBACK
```
Comment (pan3793) [inline sbin/load-celeborn-env.sh:80]:
```suggestion
    if [ -f "$JEMALLOC_PATH" ]; then
      export LD_PRELOAD="$LD_PRELOAD:$JEMALLOC_PATH"
    elif [ -f "$JEMALLOC_FALLBACK" ]; then
      export LD_PRELOAD="$LD_PRELOAD:$JEMALLOC_FALLBACK"
```

**Review by pan3793 — APPROVED** [review]

**Comment by waitinfuture** [comment]
Merging to main/0.3

File: sbin/load-celeborn-env.sh [hunk 11]
```diff
@@ -69,3 +69,24 @@ if [ "$CELEBORN_PID_DIR" = "" ]; then
   export CELEBORN_PID_DIR="${CELEBORN_HOME}/pids"
 fi
 
+# jemalloc memory allocator is enabled by default, you may set DISABLE_JEMALLOC to true in celeborn-env.sh if it's not as you wish.
```
Comment (cxzl25) [inline sbin/load-celeborn-env.sh:72]: Not every OS has jemalloc installed, and now the start and stop commands will have multiple WARNING.

We can disable it by default, then enable it in `celeborn-env.sh`, and add instructions to the website or README.

```bash
./sbin/start-master.sh
```

```
WARNING: attempted to load jemalloc from /usr/lib/arm64-linux-gnu/libjemalloc.so and /usr/lib/x86_64-linux-gnu/libjemalloc.so but the library couldn't be found. glibc will be used instead.
WARNING: attempted to load jemalloc from /usr/lib/arm64-linux-gnu/libjemalloc.so and /usr/lib/x86_64-linux-gnu/libjemalloc.so but the library couldn't be found. glibc will be used instead.
starting org.apache.celeborn.service.deploy.master.Master
```

Comment (pan3793) [reply sbin/load-celeborn-env.sh:72]: sound reasonable. how about 

1) make `PREFER_JEMALLOC` and `JEMALLOC_PATH` configurable in `celeborn-env.sh`
2) disable `PREFER_JEMALLOC` by default, at least for `branch-0.3`, we may enable it by default in the future
3) consider the adoption rate of Linux distribution, when `PREFER_JEMALLOC=true` and `JEMALLOC_PATH` is absent, search the default installed path of jemalloc libraries on Debian and CentOS 7
